session/session.go

```diff
@@ -2604,16 +2604,194 @@ func (s *session) Auth(user *auth.UserIdentity, authentication, salt []byte) err
+				"unlimited", "unlimited")
+		}
+
+		lds := strconv.FormatInt(passwordLocking.PasswordLockTimeDays, 10)
+		return privileges.GenerateAccountAutoLockErr(passwordLocking.FailedLoginAttempts, user, host, lds, lds)
+	}
+	pm := privilege.GetPrivilegeManager(s)
+	if passwordLocking.FailedLoginCount != 0 {
+		passwordLockingJSON := pm.BuildSuccessPasswordLockingJSON(passwordLocking.FailedLoginAttempts,
+			passwordLocking.PasswordLockTimeDays)
+		if passwordLockingJSON != "" {
+			if lockingErr := s.passwordLocking(user, host, passwordLockingJSON); lockingErr != nil {
+				if rollBackErr := failedLoginTrackingRollback(s); rollBackErr != nil {
+					return rollBackErr
+				}
+				return lockingErr
+			}
+		}
+	}
+	return failedLoginTrackingCommit(s)
+}
+
+func authFailedTracking(s *session, user string, host string) error {
+	passwordLocking, getErr := getFailedLoginCount(s, user, host)
+	if getErr != nil {
+		if rollBackErr := failedLoginTrackingRollback(s); rollBackErr != nil {
+			return rollBackErr
+		}
+		return getErr
+	}
+	if err := userAutoAccountLocked(s, user, host, passwordLocking.FailedLoginCount+1,
+		passwordLocking.FailedLoginAttempts, passwordLocking.PasswordLockTimeDays, passwordLocking.AutoAccountLocked); err != nil {
+		if rollBackErr := failedLoginTrackingRollback(s); rollBackErr != nil {
+			return rollBackErr
+		}
+		return err
+	}
+	if commitErr := failedLoginTrackingCommit(s); commitErr != nil {
+		return commitErr
+	}
+	return domain.GetDomain(s).NotifyUpdatePrivilege()
+}
+
+func (s *session) passwordLocking(user string, host string, newAttributesStr string) error {
+	sql := new(strings.Builder)
+	sqlexec.MustFormatSQL(sql, "UPDATE %n.%n SET ", mysql.SystemDB, mysql.UserTable)
+	sqlexec.MustFormatSQL(sql, "user_attributes=json_merge_patch(coalesce(user_attributes, '{}'), %?)", newAttributesStr)
+	sqlexec.MustFormatSQL(sql, " WHERE Host=%? and User=%?;", host, user)
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnPrivilege)
+	_, err := s.ExecuteInternal(ctx, sql.String())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func failedLoginTrackingCommit(s *session) error {
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnPrivilege)
+	_, err := s.ExecuteInternal(ctx, "COMMIT")
+	if err != nil {
+		_, rollBackErr := s.ExecuteInternal(ctx, "ROLLBACK")
+		if rollBackErr != nil {
+			return rollBackErr
+		}
+		return err
+	}
+	return nil
+}
+
+func failedLoginTrackingRollback(s *session) error {
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnPrivilege)
+	_, err := s.ExecuteInternal(ctx, "ROLLBACK")
+	return err
+}
+
+func getFailedLoginCount(s *session, user string, host string) (privileges.PasswordLocking, error) {
+	passwordLocking := privileges.PasswordLocking{}
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnPrivilege)
+	_, err := s.ExecuteInternal(ctx, "BEGIN PESSIMISTIC")
```

Comment:
Here we start an explicit transaction, but the 'COMMIT'/'ROLLBACK' code is placed in a lot of other places.

My concern is the code is vulnerable, could you improve it with something like:

```
Execute('BEGIN PESSIMISITC')
...
err, ... := authSuccessClearCount() / authFailedTracking()
if err == nil {
    Execute('COMMIT')
} else {
    Execute('ROLLBACK')
}
```
Is this possible?

Reply:
done